index.js

```diff
@@ -327,71 +318,131 @@ if(!helpers.isTestEnvironment()){
+        } else {
+            helpers.log(`Bad request, parameters missing ${JSON.stringify(validationErrors)}`)
+            res.status(400).send()
         }
+    } catch (err) {
+        helpers.log(err)
+        res.status(500).send()
+    }
+})
+
+app.post('/api/door', Validator.body(['coreid', 'data.*.deviceid', 'data.*.data', 'data.*.control']).exists(), async(request, response) => {
+    try {
+        const validationErrors = Validator.validationResult(request)
+        if(validationErrors.isEmpty()){
+            const coreId = request.body.coreid
+            const locationid = await db.getLocationIDFromParticleCoreID(coreId)
+
+            if(!locationid){
+                helpers.log(`Error - no location matches the coreID ${coreId}`)
+                response.status(400).json('No location for CoreID')
+            }
+            else{
+                const message = JSON.parse(request.body.data)
+                const signal = message.data
+                var doorSignal
+                if (signal==IM21_DOOR_STATUS.OPEN){
+                    doorSignal = SESSIONSTATE_DOOR.OPEN
+                }
+                else if (signal==IM21_DOOR_STATUS.CLOSED){
+                    doorSignal = SESSIONSTATE_DOOR.CLOSED
+                }
+                else if (signal==IM21_DOOR_STATUS.LOW_BATT){
+                    doorSignal = "LowBatt"
+                }
+                else if (signal==IM21_DOOR_STATUS.HEARTBEAT_OPEN || signal==IM21_DOOR_STATUS.HEARTBEAT_CLOSED){
+                    doorSignal = "HeartBeat"
+                }
       
-        await redis.addIM21DoorSensorData(locationid, doorSignal)
-        await handleSensorRequest(locationid)
-        response.status(200).json("OK")
+                await redis.addIM21DoorSensorData(locationid, doorSignal)
+                await handleSensorRequest(locationid)
+                response.status(200).json("OK")
+            }
+        } else {
+            helpers.log(`Bad request, parameters missing ${JSON.stringify(validationErrors)}`)
+            response.status(400).send()
+        }
+    } catch (err) {
+        helpers.log(err)
+        response.status(500).send()
     }
 })
 
 // Handler for device vitals such as wifi strength
-app.post('/api/devicevitals', async(request, response) => {
-    const coreId = request.body.coreid
-    const locationid = await db.getLocationIDFromParticleCoreID(coreId)
-    if(!locationid){
-        helpers.log(`Error - no location matches the coreID ${coreId}`)
-        response.status(400).json('No location for CoreID')
-    }else {
-        const data = JSON.parse(request.body.data)
-        const signalStrength = data.device.network.signal.strength
-        const cloudDisconnects = data.device.cloud.disconnects
-
-        redis.addVitals(locationid, signalStrength, cloudDisconnects)
-        response.status(200).json("OK")
-
+app.post('/api/devicevitals', Validator.body(['coreid']).exists(), async(request, response) => {
```

Comment:
Looks like the Validator could also check for `data`, `data.device.network.signal.strength` and `data.device.cloud.disconnects`. Or was this left out on purpose?

Reply:
Not purposefully left out that I know of, @mariocimet only mentioned checking inside data for the purpose of the one route where I have `data.*.thing` in it but that's easy to add!

Reply:
In this case would I need to explicitly check for data as if it didn't exist the other two wouldn't either?

Reply:
mm ok, I am looking at this too and those fields aren't included in the tests so it would make tests fail and I'm not sure what the values of them are supposed to be to be valid if I were to add them 🤔

from `serverTest.js`:

```
it('should return 400 to a device vitals signal with an unregistered coreID', async () => {
  let response = await chai.request(server).post('/api/devicevitals').send({coreid: 'unregisteredID',data:{deviceid:'5A:2B:3C',data: 'closed',control: 'AA'}})
  expect(response).to.have.status(400)
})
```

Reply:
To find out some valid values, you could try asking @sampathsatti . Or just try it out - you could add a log message, deploy to dev, and then watch what comes in.

Reply:
It would also be a good opportunity to write some tests that will fail when those values aren't provided.

Reply:
yeah it's just data.data

Reply:
those values are in the IM21 Door Status enum, we already vary it in the tests using the enum for open vs. closed signals

`'use strict'

const IM21_DOOR_STATE = {
    OPEN: '02',
    CLOSED: '00',
    HEARTBEAT_OPEN: '0A',
    HEARTBEAT_CLOSED: '08',
    LOW_BATT: '04'
}`